util/chunk/disk.go

```diff
@@ -164,16 +169,16 @@ func (l *ListInDisk) GetChunk(chkIdx int) (*Chunk, error) {
 
 // GetRow gets a Row from the ListInDisk by RowPtr.
 func (l *ListInDisk) GetRow(ptr RowPtr) (row Row, err error) {
-	err = l.flush()
 	if err != nil {
 		return
 	}
 	off := l.offsets[ptr.ChkIdx][ptr.RowIdx]
 	var underlying io.ReaderAt = l.disk
 	if l.ctrCipher != nil {
-		underlying = encrypt.NewReader(l.disk, l.ctrCipher)
+		underlying = NewReaderWithCache(encrypt.NewReader(l.disk, l.ctrCipher), l.cipherWriter.GetCache(), l.cipherWriter.GetCacheDataOffset())
 	}
-	r := io.NewSectionReader(checksum.NewReader(underlying), off, l.offWrite-off)
+	checksumReader := NewReaderWithCache(checksum.NewReader(underlying), l.checksumWriter.GetCache(), l.checksumWriter.GetCacheDataOffset())
+	r := io.NewSectionReader(checksumReader, off, l.offWrite-off)
```

Comment:
This question really makes sense. Especially in the previous usage scenarios: there is no writing after reading. So the every time allocation is unnecessary.

I add a new member `isReaderStale`. After writing, `isReaderStale` will be true, and when reading, we will create a new reader. If `isReaderStale` is false, we will not create new reader.

Reply:
> This question really makes sense. Especially in the previous usage scenarios: there is no writing after reading. So the every time allocation is unnecessary.
> 
> I add a new member `isReaderStale`. After writing, `isReaderStale` will be true, and when reading, we will create a new reader. If `isReaderStale` is false, we will not create new reader.

This approch may cause data race. Because we update `l.r` in `GetRow()`, that's a write, there maybe a read of `l.r` at the same time. We can create a new reader in `Add()`, but it still needs memory allocation. So I can't think of the best of both worlds. I will indicate this on PR description.